### PR TITLE
Refactor to track meshmode array context changes

### PR DESCRIPTION
**Pull request by inducer**
- Force device scalars in tests and examples
- Use meshmode transform metadata
- Give grudge its own array context, use it

@alexfikl, could you take a quick look once the tests pass?

### CODE REVIEW

File: grudge/array_context.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        _PytestPyOpenCLArrayContextFactoryWithClass):
+    actx_class = PyOpenCLArrayContext
+
+
+# deprecated
+class PytestPyOpenCLArrayContextFactoryWithHostScalars(
+        _PytestPyOpenCLArrayContextFactoryWithClass):
+    actx_class = PyOpenCLArrayContext
+    force_device_scalars = False
+
```
Comment (alexfikl) [inline grudge/array_context.py:52]:
```suggestion

register_pytest_array_context_factory("grudge.pyopencl",
        PytestPyOpenCLArrayContextFactory)
```

File: grudge/array_context.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+from meshmode.array_context import PyOpenCLArrayContext as _PyOpenCLArrayContextBase
+from arraycontext.pytest import (
```
Comment (alexfikl) [inline grudge/array_context.py:29]:
```suggestion
from arraycontext.pytest import (
		register_pytest_array_context_factory,
```

File: grudge/reductions.py [hunk 3]
```diff
@@ -289,7 +300,9 @@ def _apply_elementwise_reduction(
     @memoize_in(actx, (_apply_elementwise_reduction,
                        "elementwise_%s_prg" % op_name))
     def elementwise_prg():
-        return make_loopy_program(
+        # FIXME: This computes the reduction value redundantly for each
+        # output DOF.
```
Comment (alexfikl) [inline grudge/reductions.py:304]: This looks eerily similar to the same reduction in pytential, so probably needs fixing there too.

**Comment by alexfikl** [comment]
Oops, I took a quick look and didn't notice I didn't start a review to queue up the comments. I'll give it a more serious look when the tests pass!

File: grudge/reductions.py (hunk 3 above)
Comment (inducer) [reply grudge/reductions.py:304]: https://github.com/inducer/pytential/pull/100

File: grudge/array_context.py (hunk 2 above)
Comment (inducer) [reply grudge/array_context.py:29]: Done as part of this force-push: https://github.com/inducer/grudge/compare/b8eaa66bc10f8e659f902ad11da4962b7bbef0ab..335d27cfc84f222795693d8d6d90bfab21be0247

File: grudge/array_context.py (hunk 1 above)
Comment (inducer) [reply grudge/array_context.py:53]: Done as part of this force-push: https://github.com/inducer/grudge/compare/b8eaa66bc10f8e659f902ad11da4962b7bbef0ab..335d27cfc84f222795693d8d6d90bfab21be0247

File: grudge/reductions.py [hunk 4]
```diff
@@ -160,8 +159,9 @@ def nodal_sum_loc(dcoll: DiscretizationCollection, dd, vec) -> float:
         return sum(nodal_sum_loc(dcoll, dd, vec[idx])
                    for idx in np.ndindex(vec.shape))
 
+    # FIXME: do not force result onto host
     actx = vec.array_context
-    return sum([actx.np.sum(grp_ary) for grp_ary in vec])
+    return sum([actx.to_numpy(actx.np.sum(grp_ary))[()] for grp_ary in vec])
```
Comment (alexfikl) [inline grudge/reductions.py:164]: Why is this needed? I would have expected this to stay the way it was (with a `to_numpy` in `allreduce` above)

**Comment by inducer** [comment]
> Oops, I took a quick look and didn't notice I didn't start a review to queue up the comments. I'll give it a more serious look when the tests pass!

No worries at all. Also, no need for super scrutiny, I think (hope?) most of this is straightforward.

File: grudge/reductions.py [hunk 5]
```diff
@@ -197,8 +196,13 @@ def nodal_min_loc(dcoll: DiscretizationCollection, dd, vec) -> float:
                    for idx in np.ndindex(vec.shape))
 
     actx = vec.array_context
-    return reduce(lambda acc, grp_ary: actx.np.minimum(acc, actx.np.min(grp_ary)),
-                  vec, np.inf)
+
+    # FIXME: do not force result onto host
+    return reduce(
+            lambda acc, grp_ary: np.minimum(
```
Comment (alexfikl) [inline grudge/reductions.py:202]:
```suggestion
            lambda acc, grp_ary: actx.np.minimum(
```
?

**Review by alexfikl — APPROVED** [review]
Two small nitpicks and the rest looks good to me!

File: grudge/reductions.py (hunk 5 above)
Comment (inducer) [reply grudge/reductions.py]: Done in this force push: https://github.com/inducer/grudge/compare/335d27cfc84f222795693d8d6d90bfab21be0247..8131936e38f02b590261198529c8d167362b9d39

File: grudge/reductions.py (hunk 4 above)
Comment (inducer) [reply grudge/reductions.py]: * Fixed for `sum` here: https://github.com/inducer/grudge/compare/335d27cfc84f222795693d8d6d90bfab21be0247..8131936e38f02b590261198529c8d167362b9d39
* Added a better FIXME for the others in the same push
